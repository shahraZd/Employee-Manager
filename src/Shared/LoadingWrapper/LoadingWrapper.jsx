/**
 *
 * LoadingWrapper
 *
 */

import React, { memo, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Skeleton, message } from 'antd'
import { isEmpty } from 'lodash'
import UnavailableData from '../UnavailableData'
import { renderArrayErrors } from '../utils/errors.helper'
import './loading-wrapper.scss'

function LoadingWrapper({
  children,
  className,
  errorTypeGet,
  errorTypePost,
  skeletonConfig,
  errorIconType,
  renderErrorUi,
  displayAlert,
}) {
  const alert = body => {
    message.error(body)
  }
  const renderGetErrors = () => {
    if (!isEmpty(errorTypeGet)) {
      const errorUi = renderErrorUi || (
        <UnavailableData message={errorTypeGet} iconType={errorIconType} />
      )
      return (
        <div className="unavailableData-holder global-flex-vertical-horizontal-center">
          {errorUi}
        </div>
      )
    }
    return children
  }

  const renderPostErrorsMessage = () => {
    if (!isEmpty(errorTypePost)) {
      return (
        <div className={`x-error-message ${className}`}>
          {renderArrayErrors(errorTypePost)}
        </div>
      )
    }
    return <Fragment />
  }
  const errorBody = () => {
    if (!isEmpty(errorTypeGet)) {
      return errorTypeGet
    }
    if (!isEmpty(errorTypePost)) {
      return errorTypePost
    }
    return []
  }

  if (displayAlert) {
    const body = errorBody()
    if (!isEmpty(body)) {
      alert(body)
    }

    return children
  }
  return (
    <Fragment>
      <Skeleton {...skeletonConfig}>{renderGetErrors()}</Skeleton>
      {renderPostErrorsMessage()}
    </Fragment>
  )
}

LoadingWrapper.propTypes = {
  renderErrorUi: PropTypes.any,
  children: PropTypes.node,
  errorTypeGet: PropTypes.any,
  errorTypePost: PropTypes.any,
  skeletonConfig: PropTypes.object,
  errorIconType: PropTypes.string,
  displayAlert: PropTypes.bool,
  className: PropTypes.string,
}
LoadingWrapper.defaultProps = {
  skeletonConfig: { loading: false },
  errorTypeGet: [],
  errorTypePost: [],
  displayAlert: false,
  children: <Fragment />,
  className: '',
}

export default memo(LoadingWrapper)
