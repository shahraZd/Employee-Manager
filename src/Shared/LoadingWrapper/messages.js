/*
 * LoadingWrapper Messages
 *
 * This contains all the text for the LoadingWrapper component.
 */

import { defineMessages } from 'react-intl'

export const scope = 'app.components.LoadingWrapper'

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the LoadingWrapper component!',
  },
})
