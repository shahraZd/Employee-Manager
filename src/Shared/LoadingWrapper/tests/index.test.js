import React, { Fragment } from 'react'
import { render } from 'react-testing-library'
import { Provider } from 'react-redux'
import { browserHistory, BrowserRouter } from 'react-router-dom'
import { IntlProvider } from 'react-intl'
import LoadingWrapper from '../index'
import { DEFAULT_LOCALE } from '../../../../i18n'
import configureStore from '../../../../configureStore'
import 'jest-prop-type-error'

const initialProps = {
  renderErrorUi: {},
  children: <Fragment />,
  errorTypeGet: [
    {
      errors: {
        fetchingStudentList: jest.fn(),
        ownCheckpointAssignmentErrors: { value: {} },
      },
    },
  ],
  errorTypePost: [
    {
      addOrUpdateHolidayErrors: [{ other: {} }],
    },
  ],
  skeletonConfig: {},
  errorIconType: '',
  displayAlert: false,

 
}

describe('<LoadingWrapper />', () => {
  let store
  beforeAll(() => {
    store = configureStore({}, browserHistory)
  })

  it('Expect to crash if propTypes failed', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <LoadingWrapper {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>,
    )
  })

  it('Expect to not log in console', () => {
    const spy = jest.spyOn(global.console, 'log')
    render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <LoadingWrapper {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>,
    )
    expect(spy).not.toHaveBeenCalled()
  })

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')
    render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <LoadingWrapper {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>,
    )
    expect(spy).not.toHaveBeenCalled()
  })

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <BrowserRouter>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <LoadingWrapper {...initialProps} />
          </IntlProvider>
        </BrowserRouter>
      </Provider>,
    )
    expect(firstChild).toMatchSnapshot()
  })
})
