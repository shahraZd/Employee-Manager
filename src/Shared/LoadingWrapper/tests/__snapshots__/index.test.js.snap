// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingWrapper /> Should render and match the snapshot 1`] = `
<div
  class="ant-skeleton"
>
  <div
    class="ant-skeleton-content"
  >
    <h3
      class="ant-skeleton-title"
      style="width: 38%;"
    />
    <ul
      class="ant-skeleton-paragraph"
    >
      <li />
      <li />
      <li
        style="width: 61%;"
      />
    </ul>
  </div>
</div>
`;
